libs/ml/include/ml/dataloaders/tensor_dataloader.hpp

```diff
@@ -113,38 +128,76 @@ template <typename LabelType, typename InputType>
 bool TensorDataLoader<LabelType, InputType>::AddData(TensorType const &data,
                                                      TensorType const &labels)
 {
-
   data_         = data.Copy();
   labels_       = labels.Copy();
-  data_cursor_  = 0;
-  label_cursor_ = 0;
-
-  n_samples_ = data_.shape().at(data_.shape().size() - 1);
+  train_cursor_ = 0;
 
   batch_label_dim_ = labels_.shape().size() - 1;
   batch_data_dim_  = data_.shape().size() - 1;
 
+  n_samples_ = data_.shape().at(batch_data_dim_);
+
+  n_validation_samples_ =
+      static_cast<SizeType>(validation_to_train_ratio_ * static_cast<float>(n_samples_));
+  n_train_samples_ = n_samples_ - n_validation_samples_;
+
+  validation_offset_ = n_train_samples_;
+
   return true;
 }
 
 template <typename LabelType, typename InputType>
 typename TensorDataLoader<LabelType, InputType>::SizeType
 TensorDataLoader<LabelType, InputType>::Size() const
 {
-  return n_samples_;
+  if (this->mode_ == DataLoaderMode::VALIDATE)
+  {
+    return n_validation_samples_;
+  }
+  else
+  {
+    return n_train_samples_;
+  }
 }
 
 template <typename LabelType, typename InputType>
 bool TensorDataLoader<LabelType, InputType>::IsDone() const
 {
-  return (data_cursor_ >= data_.shape(batch_data_dim_));
+  if (this->mode_ == DataLoaderMode::TRAIN)
```

Comment:
similarly as with GetNext() we would not need the if/else statement here either.

If you do not implement the suggested change, please raise an issue and leave a TODO in the GetNext